### DISCUSSION TITLE
Why does my driveslist detection code need `echo -n` now?

### DISCUSSION THREAD

**Discussion by ecm-pushbx** [dosemu2/dosemu2#2070] [Q&A]
I found that with today's dosemu revision, my ldosboot test script required a small change: https://hg.pushbx.org/ecm/ldosboot/rev/007ddb5647c7

In particular, this code:

```
    driveslist="$("$DOSEMU" -dumb -quiet -E lredir -te 2> /dev/null < /dev/null |
      sed -re '/^about to execute|^current drive redirections/Id' )"
    dosemu_drives="$(echo "$driveslist" | wc -l)"
```

Required the change to use the `-n` switch to the `echo` command:

```
    dosemu_drives="$(echo -n "$driveslist" | wc -l)"
```

Without this change, `$dosemu_drives` is set to 3 even though dosemu2 only displays two drive redirections (drive C: and D:).

Did something change about the line endings perhaps?

**Comment by ecm-pushbx** [comment]
(Submitted this too early, root post edited to add the missing content.)

**Comment by ecm-pushbx** [comment]
It is required on 652be6a29 on our local machine (amd64 Debian 12). On 3b250d8d0 and 4beffeff4 on the same machine it seems to not be required.

On the server (amd64 Debian 11), 9cf2b7560 and the newer 4beffeff4 do not need it.

**Comment by stsp** [comment]
Should now be fixed (in comcom32).

**Reply by ecm-pushbx** [reply]
I don't use comcom32 though.

**Comment by ecm-pushbx** [comment]
Might make this a bug report after all. The update to my repo breaks `./test.sh hdimage` on the amd64 Debian 11 server with 9cf2b7560.

**Reply by ecm-pushbx** [reply]
`DOSEMU=~/local.new/bin/dosemu ./test.sh hdimage direct` as well, on the newer 4beffef

**Comment by stsp** [comment]
What's the use of making a bug or
pointing to various unrelated commits?
You have a trivial script where a variable
content mismatches.
Can't you find out yourself the difference
in the content of that var?

**Reply by ecm-pushbx** [reply]
Good idea. It's the fix to https://github.com/dosemu2/dosemu2/discussions/2061 that appears to cause the error:

```
ldosboot/test$ grep podhex ~/.bashrc 
function podhex() { od -Ax -tx1z "$@" | perl -pe 's/^([0-9A-Fa-f]+(:? +([0-9A-Fa-f]+ +)+)?)/uc($1)/eg;s/^([0-9A-F]+(?: +[0-9A-F]+){8}) /$1-/g'; }
ldosboot/test$ driveslist="$(~/proj/dosemu2/bin/dosemu -dumb -quiet -E lredir -te | sed -re '/^about to execute|^current drive redirections/Id')" && echo "$driveslist" | podhex
000000 43 3A 20 3D 20 2F 68 6F-6D 65 2F 65 63 6D 2F 2E  >C: = /home/ecm/.<
000010 64 6F 73 65 6D 75 2F 64-72 69 76 65 5F 63 20 61  >dosemu/drive_c a<
000020 74 74 72 69 62 20 3D 20-52 45 41 44 2F 57 52 49  >ttrib = READ/WRI<
000030 54 45 0D 0A 44 3A 20 3D-20 2F 68 6F 6D 65 2F 65  >TE..D: = /home/e<
000040 63 6D 2F 70 72 6F 6A 2F-64 6F 73 65 6D 75 32 2F  >cm/proj/dosemu2/<
000050 63 6F 6D 6D 61 6E 64 73-20 61 74 74 72 69 62 20  >commands attrib <
000060 3D 20 52 45 41 44 20 4F-4E 4C 59 0D 0A 1B 5B 3F  >= READ ONLY...[?<
000070 31 30 30 36 6C 1B 5B 3F-31 30 30 33 6C 1B 5B 3F  >1006l.[?1003l.[?<
000080 31 30 30 32 6C 1B 5B 3F-31 30 30 30 6C 1B 5B 3F  >1002l.[?1000l.[?<
000090 39 6C 1B 5B 3F 31 6C 0A-                         >9l.[?1l.<
000098
```

**Reply by ecm-pushbx** [reply]
Perhaps you could write the escape sequences to stderr?

**Comment by stsp** [comment]
Should now work, thanks.

**Reply by ecm-pushbx** [reply]
Thanks, it works! I backed out my changeset in https://hg.pushbx.org/ecm/ldosboot/rev/9bebca5f658f

**Reply by ecm-pushbx** [reply]
The https://github.com/dosemu2/dosemu2/discussions/2061 behaviour did not revert, by the way. Tested with `killall -v dosemu.bin` and with running `mov ax, [0FFFFh]` with the int 0Dh handler reset to dosemu's own (crashing the dosemu).
